src/main/java/slimeknights/tconstruct/tables/inventory/table/PartBuilderContainer.java

```diff
@@ -148,4 +155,71 @@ public void onCrafting(PlayerEntity playerEntity, ItemStack output, IInventory c
   public boolean canMergeSlot(ItemStack stack, Slot slotIn) {
     return slotIn.inventory != this.craftResult && super.canMergeSlot(stack, slotIn);
   }
+
+  @OnlyIn(Dist.CLIENT)
+  public List<PartRecipe> getPartRecipeList() {
+    return this.partRecipes;
+  }
+
+  @OnlyIn(Dist.CLIENT)
+  public int getPartRecipeListSize() {
+    return this.partRecipes.size();
+  }
+
+  @OnlyIn(Dist.CLIENT)
+  public boolean hasPatternInPatternSlot() {
+    return this.patternSlot.getHasStack() && !this.partRecipes.isEmpty();
+  }
+
+  /**
+   * Returns the index of the selected recipe.
+   */
+  @OnlyIn(Dist.CLIENT)
+  public int getSelectedPartRecipe() {
+    return this.selectedRecipe.get();
+  }
+
+  /**
+   * Handles the given Button-click on the server, currently only used by enchanting. Name is for legacy.
+   */
+  @Override
+  public boolean enchantItem(PlayerEntity playerIn, int id) {
+    if (id >= 0 && id < this.partRecipes.size()) {
+      this.selectedRecipe.set(id);
+      this.updateResult();
+    }
+
+    return true;
+  }
+
+  private void updateAvailablePartRecipes(IInventory inventoryIn, ItemStack stack) {
+    this.partRecipes.clear();
+    this.selectedRecipe.set(-1);
+    this.craftResult.setInventorySlotContents(0, ItemStack.EMPTY);
+
+    if (!stack.isEmpty()) {
+      this.partRecipes = this.world.getRecipeManager().getRecipes(TinkerTables.partRecipeType, inventoryIn, this.world);
+    }
+  }
+
+  private void updateAvailableMaterialRecipes(IInventory inventoryIn) {
+    this.materialRecipes.clear();
+
+    this.materialRecipes = this.world.getRecipeManager().getRecipes(TinkerTables.materialRecipeType, inventoryIn, this.world);
+  }
+
+  @Nullable
+  public MaterialRecipe getMaterialRecipe() {
+    for (MaterialRecipe candidate : this.materialRecipes) {
+      if (candidate.matches(this.craftMatrix, this.world)) {
+        return candidate;
+      }
+    }
+
+    return null;
+  }
+
+  public PartRecipe getPartRecipe() {
+    return this.partRecipes.get(this.selectedRecipe.get());
```

Comment:
I'd prefer to fetch the part recipe with a matches clause. The interface it gets should contain the selected pattern and the selected material. Lets us have multiple recipes for one pattern.

Reply:
Going to leave this one for now. I am considering a few different options for a way to do custom pattern recipes, but its not a priority